src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+import org.rumbledb.exceptions.OurBadException;
+import org.rumbledb.exceptions.VariableAlreadyExistsException;
+import org.rumbledb.expressions.AbstractNodeVisitor;
+import org.rumbledb.expressions.Expression;
+import org.rumbledb.expressions.Node;
+import org.rumbledb.expressions.control.TypeSwitchExpression;
+import org.rumbledb.expressions.control.TypeswitchCase;
+import org.rumbledb.expressions.flowr.CountClause;
+import org.rumbledb.expressions.flowr.ForClause;
+import org.rumbledb.expressions.flowr.GroupByClause;
+import org.rumbledb.expressions.flowr.GroupByVariableDeclaration;
+import org.rumbledb.expressions.flowr.LetClause;
+import org.rumbledb.expressions.flowr.OrderByClause;
+import org.rumbledb.expressions.flowr.OrderByClauseSortingKey;
+import org.rumbledb.expressions.flowr.ReturnClause;
+import org.rumbledb.expressions.flowr.SimpleMapExpression;
+import org.rumbledb.expressions.flowr.WhereClause;
+import org.rumbledb.expressions.module.FunctionDeclaration;
+import org.rumbledb.expressions.module.Prolog;
+import org.rumbledb.expressions.module.VariableDeclaration;
+import org.rumbledb.expressions.postfix.DynamicFunctionCallExpression;
+import org.rumbledb.expressions.postfix.PredicateExpression;
+import org.rumbledb.expressions.primary.ContextItemExpression;
+import org.rumbledb.expressions.primary.FunctionCallExpression;
+import org.rumbledb.expressions.primary.InlineFunctionExpression;
+import org.rumbledb.expressions.primary.NamedFunctionReferenceExpression;
+import org.rumbledb.expressions.primary.VariableReferenceExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpressionVar;
+
+
+/**
+ * Dynamic context visitor. Populates the dynamic context to evaluate the main expression.
+ */
+public class VariableDependenciesVisitor extends AbstractNodeVisitor<Void> {
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    Map<Node, Set<String>> inputVariableDependencies;
+    Map<Node, Set<String>> outputVariableDependencies;
+
+    private void addInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void removeInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        for (String v : variables) {
+            this.inputVariableDependencies.get(node).remove(v);
+        }
+    }
```

Comment:
I had a few comments in this section going back and forth with certain misunderstandings. They are removed now but you have the notifications in your inbox. I have collected my results in this comment.

After checking the usage of variables in visitProlog, I have noticed that you are using the null dependencies of a node to throw a developer safeguard exceptions against non-visited cases. I am fond of this practise to increase developer efficiency and reduce bugs. However, in the current state, I believe the logic of initializing dependencies to an emptySet and updating(add/remove) them are bit too entangled. If I am interpreting this correctly, it reminds me of our discussion regarding the staticContextVisitor and executionMode usage where my initial implementation was similarly entangled in terms of initialization and updates.

How about we implement separate init methods that simply set an emptySet as dependencies for a node. The add/remove methods expect the lists to be initialized and no checks are needed since having a non-initialized node would create NullExceptions during development. 

The add remove methods can now be simplified as: 
- if variables argument is null, we can simply return
- no need to check the dependencies maps for the existence of 'node'
- simply perform the add or remove operation with parameters
- non-existence of node in the map should cause a null pointer exception which we can catch as an OurBadException as dependencies are not initialized. As such, we can remove the null-checks from where you evaluate the dependencies in visitProlog and simplify the code there.

Reply:
Very useful comments. I will address them. Thank you!

Reply:
With getters, the OurBadExceptions were now removed.

Reply:
I'm glad to hear that. I will go over the changes now